### PR TITLE
feat(palmer): register palmer.pdsi with the xarray adapter (#1016)

### LINKED ISSUE #1016: VECTORIZE-893: Register palmer.pdsi with the xarray adapter (spatial_kernel=True)
Part of the #893 performance epic. Follow-up from #937 (VECTORIZE-893: vectorize
the Palmer per-location kernel for gridded PDSI), which added a spatial block
path to `palmer.pdsi()` at the NumPy layer (`spatial_time_major=True`, see
ADR-0009) but did not register it with the xarray adapter.

## Problem

`palmer.pdsi()` now accepts a time-major `(time, *cells)` block the same way
`indices.spi`/`indices.spei` do, but `xarray_adapter.py`/`typed_public_api.py`
have no Palmer entry point at all (per ADR-0001, Palmer is the documented
exception requiring a separate decision before it gets an xarray path; ADR-0009
grants that permission without yet spending it).

## Tasks

- [ ] Register `palmer.pdsi` with `typed_public_api.py`/`xarray_adapter.py`,
  `spatial_kernel=True`, following the pattern `indices.spi`/`indices.eddi` use.
- [ ] AWC is a per-cell scalar with no time axis, structurally the same problem
  PET's latitude solves (`xarray_adapter.py`'s `_spatial_kernel_latitude`,
  `_pet_with_copy` vs `_pet_block`); copy that pattern rather than
  `_wrapped_spi`'s.
- [ ] Add the no-loop guard test in `tests/test_spatial_kernel.py`, following
  `TestSpatialKernelSkipsPerCellLoop`'s convention (monkeypatch an inner seam,
  assert call count/shape, not `== cell_count`).
- [ ] CF metadata registry entries for the Palmer outputs (`pdsi`, `phdi`,
  `pmdi`, `z_index`).
- [ ] Decide whether `palmer.scpdsi()` gets a per-cell (`vectorize=True`) xarray
  path even though it stays off the spatial block path (ADR-0009) — scPDSI has
  no xarray entry point at all today either.

## Related

- #937 — the NumPy-layer spatial block path this registers.
- ADR-0009 — the decision this follows.
- ADR-0001 / ADR-0008 — the dual-API and spatial-block-declaration contracts this must satisfy.

### PR DESCRIPTION
**Pull request by monocongo**
## Summary

Completes the Palmer xarray registration that #937 deferred. `climate_indices.pdsi()`
accepts DataArrays and returns an `xr.Dataset` with one variable per standard
Palmer index (`pdsi`, `phdi`, `pmdi`, `z_index`), each carrying its own
`CF_METADATA` entry, version, and history. A 3-D or higher input reaches
`palmer.pdsi()` once per Dask spatial block with `spatial_time_major=True`, so the
recursion runs once per block instead of once per grid cell; 1-D and 2-D inputs
keep the per-cell path.

The adapter follows the PET entry points (`xr.apply_ufunc` owned by the index)
rather than the `xarray_adapter` decorator, for the two reasons the ticket names:
AWC is a per-cell broadcast input with no time axis (the PET latitude problem),
and the kernel returns four outputs. The shared `_spatial_kernel_latitude` helper
is generalized to `_spatial_kernel_cell_param` so latitude and AWC share one
block/transposition decision.

Closes #1016.

## Scope decisions

- **scPDSI gets no xarray path (decided, not deferred).** The ticket asks whether
  `palmer.scpdsi()` should get a per-cell `vectorize=True` path. It should not, for
  the reasons ADR-0011 already records: the per-location duration-factor fit and
  Wells backtracking run four times per cell and raise `ConvergenceError`, so a
  per-cell `apply_ufunc` would only add a wrapper around a call the caller can
  already write from NumPy values. `palmer.scpdsi()` still explicitly rejects 3+-D
  input and has no adapter entry point.
- **AWC validation** is limited to what the kernel already enforces: a scalar or a
  DataArray whose dimensions are a subset of the precipitation's cell dimensions.
  An AWC carrying a dimension the index inputs lack falls back to the per-cell
  path, exactly as PET latitude does.
- **Calendar contract**: the xarray path validates monthly periodicity and a
  January start before calling the kernel, and requires a single-chunk `time`
  dimension for Dask input, matching the other spatial-kernel entry points.

## Verification

- `uv run ruff check src/ tests/` / `uv run ruff format --check src/ tests/` — clean
- `uv run mypy src/` — clean
- `uv run pytest` — 2180 passed, 423 deselected
- `uv run pytest tests/test_release_integrity.py` — 32 passed
- New coverage in `tests/test_spatial_kernel.py::TestSpatialPalmerKernel`: the
  one-call-per-block guard against `_calculate_pdsi_prepared` (6 cells reach the
  recursion once), the 2-D per-cell fallback, block-vs-pointwise bit-for-bit
  equivalence, all-missing-cell masking inside a live block, partial final year,
  Dask-block equivalence, and the NumPy five-tuple passthrough.

## Summary by Sourcery

Register the standard Palmer PDSI family with the xarray API while preserving the existing NumPy interface and documenting scPDSI's NumPy-only scope.

New Features:
- Add xarray support to `climate_indices.pdsi()`, returning a Dataset containing PDSI, PHDI, PMDI, and Z-Index outputs with coordinates, provenance, and CF metadata.

Bug Fixes:
- Reject mismatched non-time coordinates during multi-input alignment instead of silently dropping non-overlapping grid cells.

Enhancements:
- Enable gridded PDSI calculations to process one spatial block at a time while retaining per-cell behavior for lower-dimensional inputs and stable NumPy passthrough behavior.
- Generalize broadcast cell-parameter handling for shared PET latitude and Palmer AWC inputs.
- Keep scPDSI explicitly NumPy-only in accordance with its per-location computation model.

Documentation:
- Update xarray compatibility, migration, algorithm, benchmark, and architecture documentation to describe the PDSI adapter and its spatial and calendar constraints.

Tests:
- Add coverage for PDSI Dataset outputs, block-level invocation, pointwise equivalence, Dask behavior, missing cells, partial years, validation errors, coordinate mismatches, and NumPy compatibility.

## Summary by CodeRabbit

* **New Features**
  * Added a public PDSI API supporting NumPy and xarray inputs.
  * xarray results include four standard PDSI outputs with CF metadata.
  * Added spatial grid and Dask-block processing support.
  * Added validation for mismatched shared coordinates and invalid time or AWC inputs.

* **Documentation**
  * Updated compatibility guides, migration documentation, algorithm references, and ADRs to describe PDSI xarray support.
  * Clarified that scPDSI remains NumPy-only.

### CODE REVIEW

File: src/climate_indices/typed_public_api.py [hunk 1]
```diff
@@ -576,3 +583,103 @@ def eddi(
...
+
+def pdsi(
+    precips: npt.NDArray[np.float64] | xr.DataArray,
+    pet: npt.NDArray[np.float64] | xr.DataArray,
+    awc: float | npt.NDArray[np.float64] | xr.DataArray,
+    data_start_year: int | None = None,
+    calibration_year_initial: int | None = None,
+    calibration_year_final: int | None = None,
+    fitting_params: dict[str, Any] | None = None,
+    time_dim: str = "time",
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/typed_public_api.py]: Addressed in a1a33971. `spatial_time_major` is now on both `climate_indices.pdsi()` overloads and forwarded through `xarray_adapter.palmer_pdsi()` to `palmer.pdsi()` on the NumPy path, so the ADR-0009 escape hatch for an ambiguous `(time, 12, *cells)` block is reachable through the public entry point. The xarray path is unchanged: it already declares its own blocks internally, and the parameter is documented as NumPy-only. Test: `tests/test_spatial_kernel.py::TestSpatialPalmerKernel::test_pdsi_numpy_ambiguous_block_needs_declaration` asserts the undeclared block still raises the ambiguity `ValueError` and that the declared block matches the per-cell `palmer.pdsi()` results.

File: src/climate_indices/xarray_adapter.py [hunk 2]
```diff
@@ -2283,3 +2287,247 @@ def _hargreaves_block(tmin: np.ndarray, tmax: np.ndarray, tmean: np.ndarray, lat
...
+    validate_time_monotonicity(precips_da[time_dim])
+    validate_time_monotonicity(pet_da[time_dim])
+
+    # the Palmer recursion is monthly and indexes calendar months from January
+    _build_daily_calendar_plan(precips_da[time_dim], compute.Periodicity.monthly)
+    _build_daily_calendar_plan(pet_da[time_dim], compute.Periodicity.monthly)
+
+    aligned_precips, aligned_secondaries = _align_inputs(precips_da, {"pet": pet_da}, time_dim)
+    precips_da = aligned_precips
+    pet_da = aligned_secondaries["pet"]
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/xarray_adapter.py]: Addressed in a1a33971. `_align_inputs()` now verifies every non-time dimension shared by more than one input has matching coordinates (order aside) before `xr.align()` runs, and raises `CoordinateValidationError` with `coordinate_name` set to the dimension and `reason="mismatched_non_time_coordinates"` instead of silently intersecting the grids. Time is still trimmed with the existing warning. Since the helper is shared, the silent loss is closed for SPEI and every other multi-input adapter, not just Palmer. Regression test: `tests/test_spatial_kernel.py::TestSpatialPalmerKernel::test_pdsi_partially_overlapping_grid_rejected` (PET lat `[20.0, 30.0, 40.0]` against precipitation lat `[10.0, 20.0, 30.0]`).

File: src/climate_indices/xarray_adapter.py [hunk 3]
```diff
@@ -2283,3 +2287,247 @@ def _hargreaves_block(tmin: np.ndarray, tmax: np.ndarray, tmean: np.ndarray, lat
...
+                "awc must be a scalar or numpy array when precips is a numpy array. "
+                f"Got xr.DataArray with dims={awc.dims}. Use a scalar awc or convert "
+                "precips and pet to xr.DataArray for spatial broadcasting."
+            )
+        if data_start_year is None or calibration_year_initial is None or calibration_year_final is None:
+            raise ValueError(
+                "data_start_year, calibration_year_initial, and calibration_year_final are required for numpy inputs"
+            )
+        assert isinstance(precips, np.ndarray)
+        assert isinstance(pet, np.ndarray)
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/xarray_adapter.py]: Addressed in a1a33971. The `np.ndarray` asserts are replaced with `np.asanyarray()` coercion before delegating to `palmer.pdsi()`. `asanyarray` rather than `asarray` so a masked array keeps its mask (a `MaskedArray` is accepted input per `detect_input_type()`), and coercion rather than explicit rejection because `detect_input_type()` routes list, tuple, and scalar input to the NumPy path as NumPy-coercible. Test: `tests/test_spatial_kernel.py::TestSpatialPalmerKernel::test_pdsi_numpy_list_input_coerced`, which feeds nested lists and compares against the ndarray result.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/1038 (branch `feature/1016-palmer-xarray-adapter`, reviewed at `01a24d80`)
- **Lanes run:** 5 first-pass lanes (correctness/edge cases, public API contracts, xarray dataflow, test quality, docs/references) plus 3 re-validation lanes (correctness, tests, docs) after the fix pass — 8 lane runs total.
- **Validation:** `uv run ruff check src/ tests/` — passed; `uv run ruff format --check src/ tests/` — 96 files formatted; `uv run mypy src/` — no issues (28 files); `uv run pytest` — 2195 passed, 423 deselected; `uv run pytest tests/test_release_integrity.py` — 32 passed; llms regeneration verified byte-for-byte against `scripts/generate_llms_txt.py` (`tests/test_review_scripts.py` passed).
- **Fix commits:**
  - [`e894fddc` — feat(palmer): register palmer.pdsi with the xarray adapter (#1016)](https://github.com/monocongo/climate_indices/commit/e894fddc)
  - [`803e9302` — test(palmer): split the composite tuple-element assertions](https://github.com/monocongo/climate_indices/commit/803e9302)
  - [`a1a33971` — fix(palmer): reject mismatched cell grids and forward the spatial declaration](https://github.com/monocongo/climate_indices/commit/a1a33971)
  - [`b9c09ceb` — fix(palmer): harden the pdsi adapter review findings (#1016)](https://github.com/monocongo/climate_indices/commit/b9c09ceb)
  - [`df05078e` — fix(palmer): record fitting_params provenance and cover per-cell AWC pairing (#1016)](https://github.com/monocongo/climate_indices/commit/df05078e)

### Findings and fix status

#### [P2] Spatial alignment could silently drop cells from a mismatched secondary grid — Fixed
- **Location:** `src/climate_indices/xarray_adapter.py` (`_align_inputs`)
- **Origin:** Qodo review (maintainer-applied), not one of the five agent lanes
- **Problem:** The inner join aligned every shared dimension, not just `time`, so a PET field on a different spatial grid silently lost the cells the grids did not share; the alignment warning only measured the time dimension.
- **Applied fix:** Inputs sharing a non-time dimension must now carry matching coordinates (order aside) or alignment raises `CoordinateValidationError` with `reason="mismatched_non_time_coordinates"`, commit `a1a33971`.

#### [P3] Declared `(time, 12, *cells)` NumPy block was unreachable through the new facade — Fixed
- **Location:** `src/climate_indices/xarray_adapter.py` (`palmer_pdsi`, NumPy branch), `src/climate_indices/typed_public_api.py` (`pdsi` overloads)
- **Origin:** Correctness + API contracts lanes
- **Problem:** The kernel's error tells callers to declare the ambiguous block with `spatial_time_major=True`, but that keyword existed only on `palmer.pdsi`; the facade advertised the block and could not satisfy the remediation.
- **Applied fix:** `spatial_time_major` is now on both overloads, forwarded to the NumPy branch (`a1a33971`); the docstrings state the requirement instead of denying the keyword (`b9c09ceb`).

#### [P3] All-missing `float32` input leaked its dtype — Fixed
- **Location:** `src/climate_indices/xarray_adapter.py` (`_pdsi_block`, `_pdsi_per_cell`)
- **Origin:** Correctness lane
- **Problem:** The all-missing early return hands back the input array, so a fully-NaN `float32` grid produced `float32` in memory but `float64` under Dask, contradicting `output_dtypes=[float]`.
- **Applied fix:** Both kernel wrappers cast to `float` before xarray wraps the outputs; `test_pdsi_all_missing_dtype_is_float64` pins it (`b9c09ceb`).

#### [P3] AWC carrying the time dimension failed with an internal apply_ufunc error — Fixed
- **Location:** `src/climate_indices/xarray_adapter.py` (`palmer_pdsi`, xarray branch)
- **Origin:** Correctness lane
- **Problem:** A time-indexed AWC reached `xr.apply_ufunc` and raised "unexpected dimensions ['time']" without naming `awc`.
- **Applied fix:** Explicit pre-check raises `TypeError` naming `awc` and the time dimension; covered by `test_pdsi_rejects_contract_violations` (`b9c09ceb`).

#### [P3] AWC and `fitting_params` were absent from output provenance; history read `Z_INDEX` — Fixed
- **Location:** `src/climate_indices/xarray_adapter.py` (`palmer_pdsi` metadata block)
- **Origin:** Correctness + API contracts lanes
- **Problem:** The output Dataset documented none of the broadcast inputs, and the per-variable history entry used `name.upper()`, producing `Z_INDEX`; the PET wrappers record their latitude analogously.
- **Applied fix:** `awc` (scalar/ndarray/DataArray description) and a JSON-safe `fitting_params` key-set summary are recorded in attrs and history, and display names (`PDSI`, `PHDI`, `PMDI`, `Z-Index`) are used (`b9c09ceb`, `df05078e`).

#### [P3] `Raises` section named the wrong exception for mixed inputs — Fixed
- **Location:** `src/climate_indices/xarray_adapter.py` (`palmer_pdsi` docstring)
- **Origin:** API contracts lane
- **Problem:** The docstring promised `InputTypeError` for mixed NumPy/DataArray inputs; the code raises `TypeError`, which the next bullet already documented.
- **Applied fix:** The `Raises` entries now name the actual conditions and types (`b9c09ceb`).

#### [P3] Stale "Palmer has no xarray adapter" claims across docs — Fixed
- **Location:** `docs/adr/0001`, `0004`, `0008`, `0009`, `0011`, `docs/xarray_migration.md`, `docs/algorithm_refs/palmer.md`, `benchmarks/README.md`
- **Origin:** Docs lane
- **Problem:** Multiple decision records and guides still asserted a NumPy-only Palmer API or omitted `pdsi` from the xarray function list; ADR-0011 also quoted an ADR-0009 sentence deleted in #988.
- **Applied fix:** All claims now describe the PDSI-only adapter and scPDSI's NumPy-only status; the migration guide lists `pdsi` (and the previously omitted `eddi`/`percentage_of_normal`); llms bundles regenerated (`b9c09ceb`; compatibility table also updated in `a1a33971`).

#### [P3] Test gaps around the AWC broadcast seam and error paths — Fixed
- **Location:** `tests/test_spatial_kernel.py::TestSpatialPalmerKernel` (lines 1889–2060)
- **Origin:** Tests lane
- **Problem:** A uniform AWC made the per-cell broadcast assertions blind, the Dask test could pass if laziness were lost, no validation guard was exercised through the public entry point, one assertion was near-vacuous, and the dtype contract was unpinned.
- **Applied fix:** Distinct per-cell AWCs plus per-latitude pairing checks, a `chunks is not None` laziness assertion, `test_pdsi_rejects_contract_violations` (non-January, non-monthly, multi-chunk time, mixed types, time-AWC, numpy-AWC), `.all()` instead of `.any()`, and the dtype pin (`b9c09ceb`, `df05078e`, plus `803e9302`).

#### [P4] Dataset `z_index` vs CLI `zindex` name mapping undocumented — Fixed
- **Location:** `src/climate_indices/xarray_adapter.py` (`palmer_pdsi` Notes)
- **Origin:** API contracts lane
- **Problem:** The adapter names the Z-Index variable `z_index` (matching the CF registry) while the CLI NetCDF writer uses `zindex`, with nothing telling users how the two relate.
- **Applied fix:** A Notes bullet documents the shared variable name and the CLI difference (`b9c09ceb`).

#### [P4] Benchmark inventory kept a stale blocker sentence and a wrong line citation — Fixed
- **Location:** `benchmarks/README.md` (Structural blockers section)
- **Origin:** Docs lane
- **Problem:** The section still claimed there was no adapter-layer Palmer call and cited `__main__.py:1409,1411`, which is PNP NetCDF naming, not `_apply_along_axis_palmers`.
- **Applied fix:** The section now records the #937 kernel and #1016 adapter, cites the real function location, and qualifies the remaining per-cell path (scPDSI; the CLI divisions input stays per-location by design) (`b9c09ceb`, `df05078e`).

### Remaining blockers
None.
